File: Spark-Learning/src/scala/com/atguigu/sparkcore/examples/sort/SecondSort.scala
package com.atguigu.sparkcore.examples.sort

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

case class SecondSortKey(val first: Int, val second: Int) extends Ordered[SecondSortKey] {
  def compare(that: SecondSortKey): Int = {
    if (this.first - that.first == 0)
      this.second - that.second
    else
      this.first - that.first
  }
}

/**
  * 二次排序问题
  * Spark中大于两列的排序都是二次排序
  * 采用封装对象的思想,对数据进行排序
  */
object SecondSort {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
    conf.setAppName("secondarySort")
    conf.setMaster("local")
    val sc = new SparkContext(conf)
    val lines = sc.textFile("./data/secondSort.txt")
    val transRDD: RDD[(SecondSortKey, String)] =
      lines.map(s => {
        (SecondSortKey(s.split(" ")(0).toInt, s.split(" ")(1).toInt), s)
      })
    transRDD.sortByKey(false).map(_._2).foreach(println)

  }
}

File: Spark-Learning/src/java/com/atguigu/spark/core/examples/SecondSort.java
package com.atguigu.spark.core.examples;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.api.java.function.VoidFunction;
import scala.Tuple2;

import java.io.Serializable;

class MySort implements Comparable<MySort> ,Serializable {
    Integer first ;
    Integer second ;

    public MySort(Integer first,Integer second){
        this.first  = first;
        this.second = second;
    }

    @Override
    public int compareTo(MySort mysort) {
        if(this.first == mysort.first){
            return -(this.second - mysort.second);
        }else{
            return -(this.first - mysort.first);
        }
    }
}

public class SecondSort {
    public static void main(String[] args) {
        SparkConf conf =new SparkConf();
        conf.setMaster("local");
        conf.setAppName("topn");
        JavaSparkContext sc = new JavaSparkContext(conf);
        JavaRDD<String> lines = sc.textFile("./data/secondSort.txt");
        JavaPairRDD<MySort, String> mapToPair = lines.mapToPair(new PairFunction<String, MySort, String>() {
            @Override
            public Tuple2<MySort, String> call(String line) throws Exception {
                MySort mysort = new MySort(Integer.valueOf(line.split(" ")[0]), Integer.valueOf(line.split(" ")[1]));
                return new Tuple2<MySort, String>(mysort, line);
            }
        });

        mapToPair.sortByKey().foreach(new VoidFunction<Tuple2<MySort, String>>() {
            @Override
            public void call(Tuple2<MySort, String> tp) throws Exception {
                System.out.println(tp._2);
            }
        });
    }
}
